ospfd/ospf_ext.c

```diff
@@ -288,38 +335,98 @@ static struct ext_itf *lookup_ext_by_instance(struct ospf_lsa *lsa)
  */
 
 /* Extended Prefix TLV - RFC7684 section 2.1 */
-static void set_ext_prefix(struct ext_itf *exti, uint8_t route_type,
-			   uint8_t flags, struct prefix_ipv4 p)
+static struct listnode *lookup_ext_prefix(const struct ext_itf *exti,
+					  const struct prefix_ipv4 *p)
 {
+	struct prefix_sid_lsa *prefix;
+	struct listnode *node;
+	bool exists = false;
+
+	for (ALL_LIST_ELEMENTS_RO(exti->lsa.prefix_sid_list, node, prefix)) {
+		if (prefix_same(&prefix->p, p)) {
+			exists = true;
+			break;
+		}
+	}
+
+	return exists ? node : NULL;
+}
+
+static struct prefix_sid_lsa *add_mod_ext_prefix(struct ext_itf *exti,
+						 uint8_t route_type,
+						 uint8_t flags,
+						 struct prefix_ipv4 *p)
+{
+	struct prefix_sid_lsa *psid = NULL;
+	struct listnode *node;
+
+	/* check if we already have this prefix in the list */
+	node = lookup_ext_prefix(exti, p);
+
+	if (node) {
+		psid = listgetdata(node);
+	} else {
+		psid = XCALLOC(MTYPE_OSPF_PSID_LSA,
+			       sizeof(struct prefix_sid_lsa));
+		memcpy(&psid->p, p, sizeof(struct prefix_ipv4));
+		psid->tlv_prefix.af = 0; /* hardcoded in the RFC currently */
+		psid->tlv_prefix.pref_length = p->prefixlen;
+		psid->tlv_prefix.address.s_addr = p->prefix.s_addr;
+		psid->instance = get_ext_pref_instance_value();
```

Comment:
This wrong. The instance value serves to compute the Opaque ID which is unique per router and per Opaque Type. Here, you allocate one Opaque LSA per prefix SID while I understand that you would convey in the same Opaque LSA all prefix SID attached to the same loopback.
From RFC7684 section 2.1:

> The OSPFv2 Extended Prefix TLV is used to advertise additional
   attributes associated with the prefix.  Multiple OSPFv2 Extended
   Prefix TLVs MAY be advertised in each OSPFv2 Extended Prefix Opaque
   LSA.  However, since the Opaque LSA type defines the flooding scope,
   the LSA flooding scope MUST satisfy the application-specific
   requirements for all the prefixes included in a single OSPFv2
   Extended Prefix Opaque LSA.